ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -124,27 +153,108 @@ def velocity_cb(self, twist_msg):
 
     def pose_cb(self, pose_msg):
         self.pose = pose_msg.pose
-        rospy.loginfo("waypoint_updater:pose_cb pose set to  %s", self.pose)
+        rospy.logdebug("waypoint_updater:pose_cb pose set to  %s", self.pose)
 
+    # Load set of waypoints from /basewaypoints into self.waypoints
+    # this should only happen once, so we unsubscribe at end
     def waypoints_cb(self, lane_msg):
         rospy.loginfo("waypoint_updater:waypoints_cb loading waypoints")
         if not self.waypoints:
+            max_velocity = 0.0
             for waypoint in lane_msg.waypoints:
                 self.waypoints.append(waypoint)
+                if max_velocity < self.get_waypoint_velocity(waypoint):
+                    max_velocity = self.get_waypoint_velocity(waypoint)
+                # end if
             rospy.loginfo("waypoint_updater:waypoints_cb %d waypoints loaded",
                           len(self.waypoints))
+            # setting max velocity based on project requirements in
+            # Waypoint Updater Node Revisited
+            self.max_velocity = max_velocity
+            rospy.loginfo("waypoint_updater:waypoints_cb max_velocity set to "
+                          " {} based on max value in waypoints."
+                          .format(self.max_velocity))
+            # now max_velocity is known, set up dynamic reconfig
+            if not self.dyn_reconf_srv:
+                self.dyn_reconf_srv = Server(DynReconfConfig, self.dyn_vars_cb)
+                rospy.loginfo("dynamic_parm server started")
+            # end if
         else:
             rospy.logerr("waypoint_updater:waypoints_cb attempt to load "
                          "waypoints when we have already loaded %d waypoints",
                          len(self.waypoints))
-
+        # end if else
         self.subs['/base_waypoints'].unregister()
         rospy.loginfo("Unregistered from /base_waypoints topic")
 
+    # Receive a msg from /traffic_waypoint about the next stop line
     def traffic_cb(self, traffic_msg):
-        # TODO: Callback for /traffic_waypoint message. Implement
-        # self.next_traffic_light_wp = traffic_msg.data
-        pass
+        if traffic_msg.data != self.next_traffic_light_wp:
+            # do we need to queue these or will I only get the next one
+            # coming up?
+            # If car is stopped do I need to ramp up velocity, then
+            # down to reach it?
+            self.next_traffic_light_wp = traffic_msg.data
```

Comment:
If we stick with the setup Udacity provided, the tl_detector node will continuously send the waypoint as -1 as long as no red light has been detected in the camera image.  Once a red light is detected in the image and the light position is judged to be in range in front of the car, it will update the waypoint value to to the waypoint index closest to the stop line position for that light.  Once the light turns green, it will set back to -1.

No need to queue the received messages as long as we have some logic to smoothly handle transitions and judge if we have enough stopping distance to the red or if we're already in the intersection and need to continue through.

The car is stopped and a red light is detected far away, I think we should ramp up the velocity to the default until reaching close enough to start the decel again.  

These kind of logic details will need a lot of development and tuning to naturally handle the various timing situations the car might see, just like human driving up to and through intersections.  At least there aren't any police in the simulator in case we blow through some red lights!

Reply:
I am looking at using states of speedup, maintain, slowdown, stopped where state changes can be triggered by achieving a goal (like max velocity or zero velocity), or by an outside event (e.g. traffic light detected).  

Any ideas what kind of acceleration / deceleration rates we can expect to achieve.  I saw information that about 5 m/s^2 is achievable when braking in any car without losing control as an extreme upper bound - we could then have tuning from performance to comfort.

At the moment, doesn't look like braking is occurring, just throttle off, so the target deceleration rate is very relaxed.

Reply:
It’s been a while so I don’t remember the numbers so clearly, but from my drivability testing experience typical driving is definitely not at constant accel/decel rates.  It would be ideal if the planner can make accel/decel profiles based on the situation like standing start accel or braking to a stop but that might be too complicated and we can discuss this more later.

Moderate level accels would be ~2m/s^2 but for Carla I think we should stick with lighter accels more like 1m/s^2 peak for now.

Reply:
My idea is to use the general linear deceleration rate that the car can handle very comfortably to determine the time/distance to use within the JMT calculations.  I should then be able to produce nice accel/decel profiles.

Coming up to lights that are red but still far away, I might add a coast state (need to figure out the decel rate that would be), and only apply greater decel (braking) when we get within a specific distance threshold if the lights haven't changed yet.